### PR TITLE
Add a middleware for authenticating users using VisualStudio

### PR DESCRIPTION
**Pull request by serber**
An ASP.NET Core middleware for authenticating users using VisualStudio (https://app.vsaex.visualstudio.com)

### CODE REVIEW

File: samples/Mvc.Client/project.json [hunk 1]
```diff
@@ -14,46 +14,47 @@
...
-    "Microsoft.AspNetCore.Server.IISIntegration": "1.0.0-*",
-    "Microsoft.AspNetCore.Server.Kestrel": "1.0.0-*",
-    "Microsoft.AspNetCore.StaticFiles": "1.0.0-*",
-    "Microsoft.Extensions.Configuration.CommandLine": "1.0.0-*",
-    "Microsoft.Extensions.Configuration.EnvironmentVariables": "1.0.0-*",
-    "Microsoft.Extensions.Logging.Console": "1.0.0-*",
-    "Microsoft.Extensions.Logging.Debug": "1.0.0-*"
-  },
+    "dependencies": {
+        "AspNet.Security.OAuth.ArcGIS": { "target": "project" },
```
Comment (PinpointTownes) [inline samples/Mvc.Client/project.json]: Please revert these style changes.

File: samples/Mvc.Client/Startup.cs [hunk 2]
```diff
@@ -43,7 +44,13 @@ public class Startup {
...
                 ClientId = "49e302895d8b09ea5656",
                 ClientSecret = "98f1bf028608901e9df91d64ee61536fe562064b",
-                Scope = { "user:email" }
+                Scope = {"user:email"}
+            });
+            
+            app.UseVisualStudioAuthentication(new VisualStudioAuthenticationOptions {
+                ClientId = "8DABAC41-0CAC-4BF6-BB79-670D566C55E8",
+                ClientSecret = "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsIng1dCI6Im9PdmN6NU1fN3AtSGpJS2xGWHo5M3VfVjBabyJ9.eyJjaWQiOiI4ZGFiYWM0MS0wY2FjLTRiZjYtYmI3OS02NzBkNTY2YzU1ZTgiLCJjc2kiOiI5NTRiZmQzNi05ZmUyLTRmMjYtOWFiYS00ZjM3MGUzOGI4MWIiLCJuYW1laWQiOiIxYTQ2NmYxYS0yZWViLTQzYWEtOTI5Yy0xMzQ5ZWFhMThjNWIiLCJpc3MiOiJhcHAudnNzcHMudmlzdWFsc3R1ZGlvLmNvbSIsImF1ZCI6ImFwcC52c3Nwcy52aXN1YWxzdHVkaW8uY29tIiwibmJmIjoxNDY2Mjg3NDY4LCJleHAiOjE2MjQwNTM4Njh9.x-waBWfWl5XqqMmvMwG7D4k0tkkfivOPlj0ZXGq__FJbU7SJ8oVBli1jJnrIX7B5D98hJe7s9FP8sUhIk4OPYOSCC5XqLYuVqr7Vo1-20uacEr9VN4nfDj5jJlw1JUtbK9g_pRLWzx9br3G1rxI4NpbWFpLtABz8wTEszC9Hlh9d1WdJK1GoyWLTu3VjTi2KIHlVqW_xv66nL9ImqX310iYT9ZOA4ljvLj_F3pK0Dc6EG6ZV5ryejvMkeA5Q8hypCwVPc2OiyIkn1CkwbjGGnZHEdRVLK2qpC2HP7FrcDj1-Tm8DlAIw5NUvvpxnmtZln5zuynPTaWyM74nsJ0b5wQ",
```
Comment (PinpointTownes) [inline samples/Mvc.Client/Startup.cs]: How is this JWT generated? Is it something you generated yourself (which is the recommended approach when using assertions for client authentication) or is it a completely static token generated by Visual Studio Online?

The JWT has an extremely long validity period (5 years), which seems rather unusual for JWT assertions... that said, I wouldn't be surprised if it was another MSFT sauce...

File: src/AspNet.Security.OAuth.VisualStudio/VisualStudioAuthenticationDefaults.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    public static class VisualStudioAuthenticationDefaults {
+        /// <summary>
+        /// Default value for <see cref="AuthenticationOptions.AuthenticationScheme"/>.
+        /// </summary>
+        public const string AuthenticationScheme = "VisualStudio";
+
+        /// <summary>
+        /// Default value for <see cref="RemoteAuthenticationOptions.DisplayName"/>.
+        /// </summary>
+        public const string DisplayName = "VisualStudio";
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.VisualStudio/VisualStudioAuthenticationDefaults.cs]: Please add a space: `Visual Studio` or `Visual Studio Online`.

File: src/AspNet.Security.OAuth.VisualStudio/project.json [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    "AspNet.Security.OAuth.Extensions": { "target": "project", "type": "build" },
+    "JetBrains.Annotations": { "type": "build", "version": "10.1.4" },
+    "Microsoft.AspNetCore.Authentication.OAuth": "1.0.0-*"
+  },
+
+  "frameworks": {
+    "net451": { },
+
+    "netstandard1.3": {
+        "dependencies": {
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.VisualStudio/project.json]: Please make the spacing consistent.

File: samples/Mvc.Client/Startup.cs (hunk 2 above)
Comment (serber) [inline samples/Mvc.Client/Startup.cs]: i create demo application for test middleware and this is apps client secret and id

Comment (PinpointTownes) [inline samples/Mvc.Client/Startup.cs]: So it's directly given by Visual Studio's portal?

Comment (serber) [inline samples/Mvc.Client/Startup.cs]: Yes

**Comment by PinpointTownes** [comment]
Please ping me when the 2 or 3 things that need to be fixed are addressed.

**Comment by PinpointTownes** [comment]
@serber pong.

**Comment by serber** [comment]
@PinpointTownes, sorry, i was busy with the wedding last month

**Comment by PinpointTownes** [comment]
> @PinpointTownes, sorry, i was busy with the wedding last month

Congratulations! :tada: 

No hurry, I just wanted to know if this PR was still active :sweat_smile:

**Comment by serber** [comment]
@PinpointTownes, i complete all your remarks.
I dont know why, but travis CI fails

**Comment by PinpointTownes** [comment]
@serber yeah, don't worry about the Travis CI failure, it's totally unrelated to your PR.

Merged: https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/commit/17f23cd1f0f30d69d0ebbc34409c58b6a750d962.

Thanks for your contribution :clap:

**Pull request closed without merging**